tests/common/cache/facts_cache.py

```diff
@@ -0,0 +1,150 @@
+            try:
+                with open(facts_file) as f:
+                    self._cache[hostname][key] = json.load(f)
+                    logger.info('Loaded cached facts "{}.{}" from {}'.format(hostname, key, facts_file))
+                    return self._cache[hostname][key]
+            except (IOError, ValueError) as e:
+                logger.error('Load json file "{}" failed with exception: {}'\
+                    .format(os.path.abspath(facts_file), repr(e)))
+                return {}
+
+    def write(self, hostname, key, value):
+        """Store facts to cache.
+
+        Args:
+            hostname (str): Hostname.
+            key (str): Name of cached facts.
+            value (obj): Value of cached facts. Usually a dictionary.
+
+        Returns:
+            boolean: Caching facts is successful or not.
+        """
+        facts_file = os.path.join(self._cache_location, '{}/{}.json'.format(hostname, key))
+        try:
+            host_folder = os.path.join(self._cache_location, hostname)
+            if not os.path.exists(host_folder):
+                logger.info('Create cache dir {}'.format(host_folder))
+                os.makedirs(host_folder)
+
+            with open(facts_file, 'w') as f:
+                json.dump(value, f, indent=2)
+                self._cache[hostname][key] = value
+                logger.info('Cached facts "{}.{}" under {}'.format(hostname, key, host_folder))
+                return True
+        except (IOError, ValueError) as e:
+            logger.error('Dump json file "{}" failed with exception: {}'.format(facts_file, repr(e)))
+            return False
+
+    def cleanup(self, hostname=None):
+        """Cleanup cached json files.
+
+        Args:
+            hostname (str, optional): Hostname. Defaults to None.
+        """
+        if hostname:
+            sub_items = os.listdir(self._cache_location)
+            if hostname in sub_items:
+                host_folder = os.path.join(self._cache_location, hostname)
+                logger.info('Clean up cached facts under "{}"'.format(host_folder))
+                shutil.rmtree(host_folder)
+            else:
+                logger.error('Sub-folder for host "{}" is not found'.format(hostname))
+        else:
+            logger.info('Clean up all cached facts under "{}"'.format(self._cache_location))
+            shutil.rmtree(self._cache_location)
+
+
+def cached(name):
+    """Decorator for enabling cache for facts.
+
+    The cached facts are to be stored by <name>.json. Because the cached json files must be stored under subfolder for
+    each host, this decorator can only be used for bound method of class which is subclass of AnsibleHostBase.
+
+    Args:
+        name ([str]): Name of the cached facts.
+
+    Returns:
+        [function]: Decorator function.
+    """
+    cache = FactsCache()
+    def decorator(target):
+        def wrapper(*args, **kwargs):
+            hostname = args[0].hostname
```

Comment:
I suggest adding a check for existence for ```hostname``` in args[0] in case this wrapper is used at none subclass of AnsibleHostBase. If ```hostname``` not exists, just call ```target``` directly.